packages/popup/src/ux-popup.ts

```diff
@@ -0,0 +1,160 @@
+    const transitionDuration = parseInt(transitionDurationString);
+    setTimeout(() => this.isWrapperOpen = false, transitionDuration);
+  }
+
+  updateAnchor() {
+    if (!this.trigger || typeof (this.trigger) === 'string') {
+      return;
+    }
+    const rect = this.trigger.getBoundingClientRect();
+    // by the time updateAnchor is called the dimensions will be known because isMeasured flag sets a class
+    const popupRect = this.element.getBoundingClientRect();
+    const triggerDistanceString = this.getVariableValue(this.element, 'popup', 'trigger-distance', UxPopupTheme.DEFAULT_TRIGGER_DISTANCE.toString());
+    const triggerDistance = parseInt(triggerDistanceString);
+    const windowEdgeDistanceString = this.getVariableValue(this.element, 'popup', 'window-edge-distance', UxPopupTheme.DEFAULT_WINDOW_EDGE_DISTANCE.toString());
+    const windowEdgeDistance = parseInt(windowEdgeDistanceString);
+    const anchor: IAnchor = { left: undefined, right: undefined, top: undefined, bottom: undefined, maxHeight: undefined, maxWidth: undefined };
+    const availableSpaceBottom = document.body.scrollTop + window.innerHeight - rect.bottom - triggerDistance - windowEdgeDistance;
+    const availableSpaceTop = rect.top - document.body.scrollTop - triggerDistance - windowEdgeDistance;
+    if (availableSpaceBottom > popupRect.height || availableSpaceBottom > availableSpaceTop) {
+      anchor.top = `${rect.top + rect.height + triggerDistance}px`;
+      anchor.maxHeight = availableSpaceBottom;
+    } else {
+      anchor.bottom = `${window.innerHeight - rect.top + triggerDistance}px`;
+      anchor.maxHeight = availableSpaceTop;
+    }
+    const availableSpaceRight = document.body.scrollLeft + window.innerWidth - rect.left - triggerDistance - windowEdgeDistance;
+    const availableSpaceLeft = rect.left - document.body.scrollLeft - triggerDistance - windowEdgeDistance;
+    if (availableSpaceRight > popupRect.width || availableSpaceRight > availableSpaceLeft) {
+      anchor.left = `${rect.left}px`;
+      anchor.maxWidth = availableSpaceRight;
+    } else {
+      anchor.right = `${window.innerWidth - rect.right}px`;
+      anchor.maxWidth = availableSpaceLeft;
+    }
+    this.anchor = anchor;
+  }
+
+  onWindowWheel(evt: Event) {
+    if (this.isOpen) {
+      if (evt.target === PLATFORM.global || !this.element.contains(evt.target as HTMLElement)) {
+        this.close();
+      }
+    }
+  }
+
+  onWindowResize() {
+    if (this.isOpen) {
+      this.updateAnchor();
+    }
+  }
+
+  onWindowClick(evt: Event) {
+    if (typeof (this.trigger) === 'string') {
+      this.trigger = DOM.querySelector(this.trigger) as HTMLElement;
+    }
+    if (this.trigger && this.trigger.contains(evt.target as HTMLElement)) {
+      this.triggerClick();
+    } else if (this.isOpen && normalizeBooleanAttribute('autoclose', this.autoclose)) {
+      this.close();
+    }
```

Comment:
I think we can avoid mutation of `trigger`, since there's no need for it

```suggestion
    const trigger= typeof this.trigger === 'string'
      ? DOM.querySelector(this.trigger)
      : this.trigger;
    if (trigger != null && trigger.contains(evt.target as HTMLEelement)) {
      this.triggerClick();
    } else if (this.isOpen && normalizeBooleanAttribute('autoclose', this.autoclose)) {
      this.close();
    }
```